apiary.apib

```diff
@@ -0,0 +1,397 @@
+FORMAT: 1A
+HOST: http://deploy.cwp.govt.nz/naut
+
+# Deploynaut API
+
+This API is a newly developed interface for the CWP Dashboard. Data structure is based on the
+[JSON API specifications](http://jsonapi.org/).
+
+
+# Using the API [/]
+
+In order to use the API, you will need to acquire a token via Service Desk request. Any request you make to
+an endpoint should include the following:
+
+## Making a request: Headers [GET]
+
+These headers will authenticate you against Deploynaut to evaluate whether you have the permissions necessary
+to access the information you are requesting. You can send a GET request to www.deploy.cwp.govt.nz/naut to
+test that you are authenticating correctly - it will return your user information.
+
++ Request (application/json)
+
+    + Headers
+
+            X-api-version: 1.0
+            X-user-email: sherbert.lemon@hogwarts.uk
+            X-user-token: fa88787fcd99787
+            Accept: application/json
+            Host: deploy.cwp.govt.nz
+
+
++ Response 200 (application/json)
+
+        {
+            "data": [
+                {
+                    "type": "users",
+                    "id": "2379",
+                    "name": "Albus Dumbledore",
+                    "email": "sherbert.lemon@hogwarts.uk",
+                    "phone": "0203040506"
+                }
+            ]
+        }
+
+
+# Stacks [/projects]
+
+## List all stacks [GET /projects]
+
+Will return a list of all stacks that you have the authority to view, with relevant datapoints.
+
++ Response 200 (application/json)
+
+        {
+            "count": 1,
+            "data": [
+                {
+                    "type": "stacks",
+                    "id": "CWP001",
+                    "name": "CWP Stack 001",
+                    "primary_domain": "mom.govt.nz",
+                    "agency": "Ministry of Magic",
+                    "size": "large",
+                    "disaster_recovery": "active",
+                    "created": "02-07-2014",
+                    "environments": {
+                        "count": 2,
+                        "data": [
+                            {
+                                "type": "environments",
+                                "id": "1",
+                                "name": "prod",
+                                "build_sha": "d0bcf80381579a529957d77a6b0a69303afebe7e",
+                                "deployer": {
```

Comment:
I think "deployer" field should be removed from here - it feels too deep if a response for "all stacks" request, includes all details of all deployers on all environments of that stack. Stig says:

> keep it slim, until we get feedback that api users often want that information. I think after some real usage you end up making utility api endpoint or providing more data so the the users doesn't need to make to many calls to get the information they want."
>
> i think you want to divide the api calls so that you use the http error codes, like 404 - no deployers found instead of trawling through the returned json... if you catch my drift. But there is that balance between easy to code and making too many api calls
> 
> if we really want to be helpful, we would provide an api link to the environment, hateos style